android/engine/src/main/java/org/smartregister/fhircore/engine/di/FhirEngineModule.kt

```diff
@@ -55,7 +55,7 @@ class FhirEngineModule {
     FhirEngineProvider.init(
       FhirEngineConfiguration(
         enableEncryptionIfSupported = !BuildConfig.DEBUG,
-        databaseErrorStrategy = DatabaseErrorStrategy.UNSPECIFIED,
+        databaseErrorStrategy = DatabaseErrorStrategy.RECREATE_AT_OPEN,
```

Comment:
> What would be the implication for using this `DatabaseErrorStrategy`?


The DatabaseErrorStrategy.RECREATE_AT_OPEN was implemented to solve a critical issue where the app crashes on startup after restoring a backup. The application is unable to read from the restored database due to this error:

`FATAL EXCEPTION: main
                                                                                                    Process: org.smartregister.opensrp.eusm, PID: 18215
                                                                                                    java.lang.RuntimeException: Unable to create application org.smartregister.fhircore.quest.QuestApplication: android.database.sqlite.SQLiteException: file is not a database: , while compiling: select count(*) from sqlite_master;
                                                                                                    	at android.app.ActivityThread.handleBindApplication(ActivityThread.java:7003)
                                                                                                    	at android.app.ActivityThread.-$$Nest$mhandleBindApplication(Unknown Source:0)
                                                                                                    	at android.app.ActivityThread$H.handleMessage(ActivityThread.java:2236)
                                                                                                    	at android.os.Handler.dispatchMessage(Handler.java:106)
                                                                                                    	at android.os.Looper.loopOnce(Looper.java:205)
                                                                                                    	at android.os.Looper.loop(Looper.java:294)
                                                                                                    	at android.app.ActivityThread.main(ActivityThread.java:8177)
                                                                                                    	at java.lang.reflect.Method.invoke(Native Method)
                                                                                                    	at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:552)
                                                                                                    	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:971)
                                                                                                    Caused by: android.database.sqlite.SQLiteException: file is not a database: , while compiling: select count(*) from sqlite_master;
                                                                                                    	at net.sqlcipher.database.SQLiteCompiledSql.native_compile(Native Method)
                                                                                                    	at net.sqlcipher.database.SQLiteCompiledSql.compile(SQLiteCompiledSql.java:89)
                                                                                                    	at net.sqlcipher.database.SQLiteCompiledSql.<init>(SQLiteCompiledSql.java:62)
                                                                                                    	at net.sqlcipher.database.SQLiteProgram.<init>(SQLiteProgram.java:91)
                                                                                                    	at net.sqlcipher.database.SQLiteQuery.<init>(SQLiteQuery.java:50)
                                                                                                    	at net.sqlcipher.database.SQLiteDirectCursorDriver.query(SQLiteDirectCursorDriver.java:60)
                                                                                                    	at net.sqlcipher.database.SQLiteDatabase.rawQueryWithFactory(SQLiteDatabase.java:2019)
                                                                                                    	at net.sqlcipher.database.SQLiteDatabase.rawQuery(SQLiteDatabase.java:1905)
                                                                                                    	at net.sqlcipher.database.SQLiteDatabase.keyDatabase(SQLiteDatabase.java:2672)
                                                                                                    	at net.sqlcipher.database.SQLiteDatabase.openDatabaseInternal(SQLiteDatabase.java:2602)
                                                                                                    	at net.sqlcipher.database.SQLiteDatabase.openDatabase(SQLiteDatabase.java:1250)
                                                                                                    	at net.sqlcipher.database.SQLiteDatabase.openOrCreateDatabase(SQLiteDatabase.java:1325)
                                                                                                    	at net.sqlcipher.database.SQLiteOpenHelper.getWritableDatabase(SQLiteOpenHelper.java:167)
                                                                                                    	at com.google.android.fhir.db.impl.SQLCipherSupportHelper.getWritableDatabase(SQLCipherSupportHelper.kt:91)
                                                                                                    	at com.google.android.fhir.db.impl.SQLCipherSupportHelper.<init>(SQLCipherSupportHelper.kt:122)
                                                                                                    	at com.google.android.fhir.db.impl.SQLCipherSupportHelper.<init>(SQLCipherSupportHelper.kt:36)
                                                                                                    	at com.google.android.fhir.db.impl.DatabaseImpl.lambda$2$lambda$1(DatabaseImpl.kt:97)
                                                                                                    	at com.google.android.fhir.db.impl.DatabaseImpl$$ExternalSyntheticLambda0.create(D8$$SyntheticClass:0)
                                                                                                    	at com.google.android.fhir.db.impl.ResourceDatabase_Impl.createOpenHelper(ResourceDatabase_Impl.java:378)
                                                                                                    	at androidx.room.RoomDatabase.init(RoomDatabase.kt:186)
                                                                                                    	at androidx.room.RoomDatabase$Builder.build(RoomDatabase.kt:1358)
                                                                                                    	at com.google.android.fhir.db.impl.DatabaseImpl.<init>(DatabaseImpl.kt:116)
                                                                                                    	at com.google.android.fhir.FhirServices$Builder.build(FhirServices.kt:83)
                                                                                                    	at com.google.android.fhir.FhirEngineProvider.getOrCreateFhirService(FhirEngineProvider.kt:99)
                                                                                                    	at com.google.android.fhir.FhirEngineProvider.getInstance(FhirEngineProvider.kt:63)
                                                                                                    	at org.smartregister.fhircore.engine.di.FhirEngineModule.provideFhirEngine(FhirEngineModule.kt:82)
                                                                                                    	at org.smartregister.fhircore.engine.di.FhirEngineModule_ProvideFhirEngineFactory.provideFhirEngine(FhirEngineModule_ProvideFhirEngineFactory.java:60)
                                                                                                    	at org.smartregister.fhircore.quest.DaggerQuestApplication_HiltComponents_SingletonC$SingletonCImpl$SwitchingProvider.get(DaggerQuestApplication_HiltComponents_SingletonC.java:1384)
                                                                                                    	at dagger.internal.DoubleCheck.get(DoubleCheck.java:47)
                                                                                                    	at org.smartregister.fhircore.quest.DaggerQuestApplication_HiltComponents_SingletonC$SingletonCImpl$SwitchingProvider.get(DaggerQuestApplication_HiltComponents_SingletonC.java:1517)
                                                                                                    	at dagger.internal.DoubleCheck.get(DoubleCheck.java:47)
2024-09-18 14:47:49.475 18215-18215 AndroidRuntime          pid-18215                            E  	at org.smartregister.fhircore.quest.DaggerQuestApplication_HiltComponents_SingletonC$SingletonCImpl.injectQuestApplication2(DaggerQuestApplication_HiltComponents_SingletonC.java:1347)
                                                                                                    	at org.smartregister.fhircore.quest.DaggerQuestApplication_HiltComponents_SingletonC$SingletonCImpl.injectQuestApplication(DaggerQuestApplication_HiltComponents_SingletonC.java:1336)
                                                                                                    	at org.smartregister.fhircore.quest.Hilt_QuestApplication.hiltInternalInject(Hilt_QuestApplication.java:52)
                                                                                                    	at org.smartregister.fhircore.quest.Hilt_QuestApplication.onCreate(Hilt_QuestApplication.java:43)
                                                                                                    	at org.smartregister.fhircore.quest.QuestApplication.onCreate(QuestApplication.kt:59)
                                                                                                    	at android.app.Instrumentation.callApplicationOnCreate(Instrumentation.java:1316)
                                                                                                    	at android.app.ActivityThread.handleBindApplication(ActivityThread.java:6998)
                                                                                                    	... 9 more`
                                                                                                    	
This strategy ensures the app remains functional by resetting the database if it can't be read.

Reply:
``This strategy is NOT respected when opening a previously unencrypted database with an encrypted configuration or vice versa. An IllegalStateException is thrown instead.``

Out of curiosity does the new version of the APK use the same encryption for the DB?

Reply:
The PR fixes the problem at hanf. Good job on it.

Reply:
> Out of curiosity does the new version of the APK use the same encryption for the DB?

No changes were made to the encryption strategy in this PR.